c++/src/kj/async-inl.h

```diff
@@ -2052,7 +2052,16 @@ PromiseCrossThreadFulfillerPair<T> newPromiseAndCrossThreadFulfiller() {
 // coroutine implementation type via a traits class parameterized by the coroutine return and
 // parameter types. We'll name our coroutine implementation `kj::_::Coroutine<T>`,
 
-namespace kj::_ { template <typename T> class Coroutine; }
+namespace kj::_ {
+
+template <typename T> class Coroutine;
+
+template <typename T>
+concept NoWaitScope = !isSameType<Decay<T>, WaitScope>();
```

Comment:
Hmm, is there any particular reason this is better-expressed as a concept rather than a constexpr function? I see you're just using it to evaluate to its boolean value in practice.

Reply:
My initial version used
```cpp
template <typename T, NoWaitScope... Args>
struct coroutine_traits<T, Args...> {
  // ...
```
before I realized that a static_assert would provide a friendlier error message.

I don't have a strong preference among the four options (concept, constexpr variable, constexpr function, or just using isSameType() directly in the static_assert).

Reply:
For reference, here is the error message from the concept-using static_assert:
```
In file included from src/kj/async-coroutine-test.c++:22:
In file included from bazel-out/k8-fastbuild/bin/src/kj/_virtual_includes/kj-async/kj/async.h:1330:
bazel-out/k8-fastbuild/bin/src/kj/_virtual_includes/kj-async/kj/async-inl.h:2074:3: error: static assertion failed: Coroutines are not allowed to accept `WaitScope` parameters.
  static_assert((::kj::_::NoWaitScope<Args> && ...),
  ^              ~~~~~~~~~~~~~~~~~~~~~~~~~~
src/kj/async-coroutine-test.c++:254:3: note: in instantiation of template class 'std::coroutine_traits<kj::Promise<void>, kj::WaitScope &>' requested here
  co_return;
  ^
bazel-out/k8-fastbuild/bin/src/kj/_virtual_includes/kj-async/kj/async-inl.h:2074:18: note: because 'kj::WaitScope &' does not satisfy 'NoWaitScope'
  static_assert((::kj::_::NoWaitScope<Args> && ...),
                 ^
bazel-out/k8-fastbuild/bin/src/kj/_virtual_includes/kj-async/kj/async-inl.h:2060:23: note: because '!isSameType<Decay<kj::WaitScope &>, kj::WaitScope>()' evaluated to false
concept NoWaitScope = !isSameType<Decay<T>, WaitScope>();
```

and here with using isSameType() directly:
```
In file included from src/kj/async-coroutine-test.c++:22:
In file included from bazel-out/k8-fastbuild/bin/src/kj/_virtual_includes/kj-async/kj/async.h:1330:
bazel-out/k8-fastbuild/bin/src/kj/_virtual_includes/kj-async/kj/async-inl.h:2069:3: error: static assertion failed due to requirement '!::kj::isSameType()': Coroutines are not allowed to accept `WaitScope` parameters.
  static_assert(!(::kj::isSameType<::kj::Decay<Args>, ::kj::WaitScope>() || ...),
  ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/kj/async-coroutine-test.c++:254:3: note: in instantiation of template class 'std::coroutine_traits<kj::Promise<void>, kj::WaitScope &>' requested here
  co_return;
  ^
```

Reply:
I think my logic goes like this:
1. If we want to define a constexpr nullary function template, e.g. `isSameType<T, U>()`, we should consider a constexpr variable template instead, because the `()` syntax doesn't really add much value. I suspect the reason we didn't use constexpr variables for this sort of thing in the past is mostly that constexpr functions arrived in the C++ standard first.
2. If we are defining a constexpr _boolean_ variable template, we should consider using a concept instead, because `template <typename T> concept ...` is the same thing as `template <typename T> constexpr bool ...`, with the added feature that the concept can be used to constrain template parameters. (`constexpr bool` can in fact be a drop-in replacement for `concept` in this particular code.)

I'll leave it as-is, because I'd like to start getting used to using them, though I admit this isn't the most compelling example.

> (I would approve this PR except it still has the co_magic define.)

D'oh, fixed.